### PR TITLE
fix(boxplot) correctly handle series.encode with category axis

### PR DESCRIPTION
**Pull request by jonoshearman**
## Brief Information

This pull request is in the type of:

- [x] bug fixing
- [ ] new feature
- [ ] others

### What does this PR do?

Fixes issue with series.encode for boxplot when category axis is used.

### Fixed issues
#20319 

### Before: What was the problem?

When a category axis is used for a boxplot series, an ordinalValue is always inserted to the series.data which breaks any encoding rules. if the category axis does not set it's own data then the axis labels will be the integer ordinalValue.

### After: How does it behave after the fixing?

Now if an encode rule is applied for the associated axis (x/y) then the ordinalValue is not inserted and the encoding rules work as expected.

## Document Info

One of the following should be checked.

- [x] This PR doesn't relate to document changes
- [ ] The document should be updated later
- [ ] The document changes have been made in apache/echarts-doc#xxx

### ZRender Changes

- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

test/boxplot-category.html added to show visual example.

### Merging options

- [ ] Please squash the commits into a single one when merging.

### CODE REVIEW

File: src/chart/helper/whiskerBoxCommon.ts [hunk 1]
```diff
@@ -52,6 +52,16 @@ class WhiskerBoxCommonMixin<Opts extends CommonOption> {
 
     defaultValueDimensions: CoordDimensionDefinition['dimsDef'];
 
+    /**
+     * @private
+     */
+    _hasEncodeRule(key: string) {
+        const encodeRules = this.getEncode();
+        if(encodeRules?.data?.has(key))
```
Comment (Ovilia) [inline src/chart/helper/whiskerBoxCommon.ts:60]: We avoid using `?.` under compatible considerations. Please use `encodeRules && encodeRules.data && encodeRules.data.has(key)` instead.

**Review by Ovilia — CHANGES_REQUESTED** [review]
Thanks for your contribution. Please fix the lint problems.

File: src/chart/helper/whiskerBoxCommon.ts (hunk 1 above)
Comment (jonoshearman) [reply src/chart/helper/whiskerBoxCommon.ts]: Thanks for the review @Ovilia
Changes pushed.

File: src/chart/helper/whiskerBoxCommon.ts [hunk 2]
```diff
@@ -52,6 +52,17 @@ class WhiskerBoxCommonMixin<Opts extends CommonOption> {
 
     defaultValueDimensions: CoordDimensionDefinition['dimsDef'];
 
+    /**
+     * @private
+     */
+    _hasEncodeRule(key: string) {
+        const encodeRules = this.getEncode();
+        if (encodeRules && encodeRules.data && encodeRules.data.has(key)) {
+            return encodeRules.data.get(key) !== null;
```
Comment (Ovilia) [inline src/chart/helper/whiskerBoxCommon.ts:61]: Using `!=` instead of `!==` please.

**Review by Ovilia — CHANGES_REQUESTED** [review]
There's just a minor issue. Everything else seems fine to me.

**Comment by Ovilia** [comment]
Please don't use forced push because it would be hard to track the changes of previous reviews.

File: src/chart/helper/whiskerBoxCommon.ts (hunk 2 above)
Comment (jonoshearman) [reply src/chart/helper/whiskerBoxCommon.ts]: Thanks, changes pushed

**Review by Ovilia — APPROVED** [review]

**Comment by Ovilia** [comment]
We sincerely appreciate your valuable contributions! We are currently running a special promotion for users who submit a Pull Request between September 25, 2024, and December 31, 2024. Eligible participants will receive a two-year enterprise membership to Baidu Comate, valued at 2000 RMB (approximately 285 USD).

To learn more about this product, please visit https://comate.baidu.com/. If you are interested in this offer, kindly send an email with the link to your Pull Request along with your Comate username and email address to ovilia@apache.org to claim your membership.

Thank you once again for your support!

**Pull request merged**